hal/src/peripherals/adc/config.rs

```diff
@@ -0,0 +1,171 @@
+use atsamd_hal_macros::hal_cfg;
+
+#[hal_cfg("adc-d5x")]
+use crate::pac::adc0;
+
+#[hal_cfg(any("adc-d21", "adc-d11"))]
+use crate::pac::adc as adc0;
+
+#[hal_cfg(any("adc-d21", "adc-d11"))]
+pub use adc0::ctrlb::Prescalerselect as Prescaler;
+
+#[hal_cfg("adc-d5x")]
+pub use adc0::ctrla::Prescalerselect as Prescaler;
+
+pub use adc0::avgctrl::Samplenumselect as SampleCount;
+
+pub use adc0::ctrlb::Resselselect as Resolution;
+
+pub use adc0::refctrl::Refselselect as Reference;
+
+/// Result accumulation strategy for the ADC
+#[derive(Copy, Clone, PartialEq, Eq)]
+pub enum Accumulation {
+    /// The ADC will read once and then the result is ready
+    Single,
+    /// The ADC will read [AdcSampleCount] samples, average them out
+    /// into a 16 bit wide value, and then the result is ready
+    Average(SampleCount),
+    /// The ADC will read [AdcSampleCount] samples, sum them
+    /// into a 16 bit wide value, and then the result is ready
+    Summed(SampleCount),
+}
+
+/// # ADC configuration builder
+///
+/// Multiple factors can affect the ADCs overall sampling rate, and this
+/// structure allows for the configuring of the majority of factors that affect
+/// the sample rate of the ADC
+///
+/// To begin with, the ADC Clock is driven by the peripheral clock divided with
+/// a divider ([AdcDivider]).
+///
+/// Each sample is read by the ADC over
+/// [AdcSettingsBuilder::sample_clock_cycles] clock cycles, and then transmitted
+/// to the ADC register over [AdcSettingsBuilder::bit_width] clock cycles (1
+/// clock cycle per bit)
+///
+/// The ADC can also be configured to combine multiple simultaneous readings in
+/// either an average or summed mode (See [AdcAccumulation]), this also affects
+/// the overall sample rate of the ADC as the ADC has to do multiple
+/// samples before a result is ready.
+///
+/// Therefore, the overall formula for calculating Sample rate (SPS) can be
+/// calculated like so:
+///
+/// ## For single sample
+/// ```
+/// SPS = (GCLK_ADC / clk_divider) / (sample_clock_cycles + bit_width)
+/// ```
+/// ## For multiple samples (Averaging or Summed)
+/// ```
+/// SPS = (GCLK_ADC / clk_divider) / (n * (sample_clock_cycles + bit_width))
+/// ```
+#[derive(Copy, Clone)]
+pub struct Config {
+    pub clk_divider: Prescaler,
+    pub sample_clock_cycles: u8,
+    pub bit_width: Resolution,
+    pub accumulation: Accumulation,
+    pub vref: Reference,
+}
+
+impl Config {
+    ///
+    /// Configure the ADC to sample at 250_000 SPS (Assuming the clock source is
+    /// 48MHz) using the following settings:
+    /// * clock divider factor of 32
+    /// * 6 clock cycles per sample
+    /// * 12bit sampling
+    /// * Single accumulation (No averaging or summing)
+    ///
+    /// ## Additional reading settings by default
+    /// * Use VDDANA as reference voltage for a full 0.0-3.3V reading
+    pub fn new() -> Self {
```

Comment:
With a little change, this design could make firmware using this API a bit easier to read, at present understanding code that uses this would require digging in to the HAL docs/source.

These could be supplied as arguments rather than magic values, or if the intent is to use a builder pattern perhaps the struct could use `Option<>` for all fields (so `new()` returns a struct with all fields `None`).

Reply:
After thinking about it, I agree that since there isn't an obvious default that  > 80% of users would use without a valid reason to deviate, we should force users to always explicitly pick their ADC parameters, UNLESS we'd want to leave the parameters to the defaults at reset.